### ISSUE TITLE
DB Table, View and edit Entry

### ISSUE THREAD

**Issue by bnomei** [k-next/plugins#60]
like suggested by @luxlogica at #plugins...

> I've been using Kirby on most of our projects for about 4 years now. The only thing that stops us from using Kirby *exclusively* is the fact that for some projects, we deal with records that number in the thousands - and then, a flat-file solution is not optimal. Some examples include an e-learning site that is estimated to have over 10,000 users, a blog with about 5,000 articles currently, and an online store which will open with just under 3,000 products. In all those use-cases, we've had to go with something other than Kirby, because we need _fast searching_ that only a database will provide. What would REALLY help us - and I believe, a lot of other developers - is if we could have a *DBTable Plugin*: that is, a Section Plugin in Kirby 3, which would provide a list-detail view to a single database table (no relations needed). The idea is that the plugin would allow us to add a section to our templates, which would allow the client to view a list of records in a database table (list view), and once a record is selected in the list, to edit it using the normal Panel fields (detail view). [...]

please contact @luxlogica at slack if interested. a budget to create this plugin seems to be available.

**Comment by bnomei** [comment]
like discussed on the slack channel maybe its enough create plugin which creates a sqllite database allowing for quicker sorts and filtering of huge list of entries? if the plugin provides a frontend to setup which fields from which template to use and attaches hooks?
like https://github.com/pedroborges/kirby-fuzzy-search just with sqlite and hooks?

**Comment by luxlogica** [comment]
@bnomei Where the current flat-file setup falls short is on both indexing/sorting, as well as searching/filtering - which are all far too slow when the number of pages reach a few thousand. This is felt in a few different places - even in the Panel. 

This is why I believed that bypassing the flat-file storage, and putting everything (not just an index of the existing page data) in the database, might be a more solid solution - but I could be wrong.

**Comment by bnomei** [comment]
@luxlogica i do not see any merrits in storing jpgs/vids of pages in an db regarding performance.
the meta data files based on the blueprint could be stored easily. problem might be if not as yaml string to be loaded but if one want to access data.
then something like a table for each fieldtype would have to be created. and a page table storing references to these. but changing blueprint definitions might make this a bit tricky.

**Comment by luxlogica** [comment]
@bnomei I was never talking about storing _image data_ in the database (don't know where that idea came from). 

I have considered several options already - such as trying to use hooks to manually keep text indexes of page fields, which could then be loaded and searched on-the-fly via javascript. I even found some ready-made javascript solutions available, but the solution they provide had lots of 'caveats' and quirks, which impact on performance, and make them difficult to maintain on the long-term. 

Consider the use-case of a online store with 3,000 products - I currently have a client in that situation. When you have 3,000 Pages with fields for title, part number, description, technical specifications - all of which are stored in various types of text fields - I've found that using the standard Kirby flat-file data storage is simply not fast enough. Even Panel operations - like searching for a specific product page - become sluggish. All we need in that situation is to be able to have a single-table database. Visitors to the shop will often search through *all* of the fields at once - ie., their search terms need to be searched through title, part number, description AND tech specs - then even a simple SQLite database starts to seem very compelling...

**Comment by bnomei** [comment]
i do not think the actual query is that slow in kirby when done in php. maybe its just building the index. @distantnative might know if with kirby 3 the index is cached between two calls of page. doing that and switching to memcache should be very fast already.
that is not a sqllite query solution - i know. but keeping the collections query chain kirby has and just bringing the index up to db speed might be a reasonable idea as well.

**Comment by jenstornell** [comment]
I have a clear idea of how it can be done. I don't want to write everything down again so here you have some reading:

https://forum.getkirby.com/t/panel-how-to-switch-file-system-to-a-database/7107

**Comment by distantnative** [comment]
@bnomei I am not sure if I understood your question correctly: There is nothing like an index cache between two different site/page loads. Was that what you asked?

**Comment by bnomei** [comment]
@distantnative yes – i guess. lets say i have a page with lots time spend on building the index for whatever reason (lots of subpages, files, multilanguage etc.).
if i view that page in browser with kirby page cache not active and hit refresh the time spent on building the index will be required again, correct?

**Comment by distantnative** [comment]
Correct.

**Comment by bnomei** [comment]
@distantnative so is there any way to not build the index again between page refreshs?

**Comment by distantnative** [comment]
@bnomei I don't think so currently, but 100% sure is only @bastianallgeier

**Comment by bnomei** [comment]
out of curiosity i want to play around a bit with data handler. @distantnative can you tell me real quickly how kirby knows its data is stored in txt files with yaml content and how to change that data handler? i only found out how they work but not how to register new ones.
https://github.com/k-next/kirby/blob/e6c980859be0c7b5d99661daceb1b139b453eed8/src/Data/Data.php
https://nnnnext.getkirby.com/docs/framework/data/handler

**Comment by bnomei** [comment]
**goal**
for some templates/models do not use default yaml handler wrting to file but use a custom ahndler defined in a plugin. this handler could but mightl not need a file (but must have one for reasons see below). more likely it will write to a sql(ite) database or get/post to an api.

**research and issues**
this is what i could find about data handlers.

- there are a handful defined as static array in kirby. but i could not find a way to add more in a clean way – like from a plugin config.
https://github.com/k-next/kirby/blob/e6c980859be0c7b5d99661daceb1b139b453eed8/src/Data/Data.php#L32
- the handler is called when the model is `save()`ing but what handler is called depends on fileextension of content file. there is no way for the model override this. so there is a one-to-one mapping from fileextension to handler (is that a good thing?).
https://github.com/k-next/kirby/blob/c96d83fa83141ea10e4e66f9394b97057507c3d0/src/Cms/ModelWithContent.php#L133
https://github.com/k-next/kirby/blob/e3899ba66cd26740a5ff38a8d224e7de0aadc344/src/Data/Data.php#L122
- in relation to the content file there is the assumption its always an existing file. which might be ok since kirby is a filebase cms afterall. but in case of an sql/api handler the file would be more or less just be empty.
https://github.com/k-next/kirby/blob/c96d83fa83141ea10e4e66f9394b97057507c3d0/src/Cms/Language.php#L127
- the content file is determined by a global option. probem is it this way ther can only be one handler for all models at one time, right? why? at least a model should be able to override this for its save function (thats possible right now but you have to recreate the save function not just set a variable at the model).
https://github.com/k-next/kirby/blob/c96d83fa83141ea10e4e66f9394b97057507c3d0/src/Cms/File.php#L227
https://github.com/k-next/kirby/blob/c96d83fa83141ea10e4e66f9394b97057507c3d0/src/Cms/App.php#L238

what i would need to take this any further would a proper example on 
- how to create and register a new handler
- how to use that handler for some but not all models
